### PR TITLE
hypre: new version 2.22.1, variant fortran, becomes AutotoolsPackage

### PR DESCRIPTION
**Pull request by cessenat**
This is the continuation from outdated PR https://github.com/spack/spack/pull/24445 with latest hypre release 2.22.1 (this release improves GPU support) and of course merge with commits performed in the meantime. This had been reviewed by @osborn9 and @alalazo.
There remains the debate, as pointed out by @adamjstewart, over Package, AutotoolsPackage versus CMakePackage as proposed by @junghans with PR https://github.com/spack/spack/pull/24817.
I personally like the autotoolspackage since I'm better used to it ! This is also the way I compile Hypre when outside of spack. As well, the simplest way to install hypre for Unix like systems is configure, as I understand from
https://github.com/hypre-space/hypre/blob/master/INSTALL.md

### CODE REVIEW

**Comment by junghans** [comment]
@cessenat one problem is that cmake's `find_package(HYPRE)` will only work if hypre was compiled with cmake as `HYPREConfig.cmake` only get generated and installed in that case. So building it with autotools will basically cut off all consumers on that side.

**Comment by cessenat** [comment]
> @cessenat one problem is that cmake's `find_package(HYPRE)` will only work if hypre was compiled with cmake as `HYPREConfig.cmake` only get generated and installed in that case. So building it with autotools will basically cut off all consumers on that side.

@junghans 
Does it mean that cmake does not work on any package that has been installed by configure ? Can't cmake find a work around ? This seems to be a big cons for cmake, not for Hypre. The thing I do not get, is that, for instance, VisIt uses cmake while Silo (which VisIt depends on) uses configure : and that works. Sorry, my knowledge of cmake is tiny...

**Comment by junghans** [comment]
> > @cessenat one problem is that cmake's `find_package(HYPRE)` will only work if hypre was compiled with cmake as `HYPREConfig.cmake` only get generated and installed in that case. So building it with autotools will basically cut off all consumers on that side.
> 
> @junghans
> Does it mean that cmake does not work on any package that has been installed by configure ? Can't cmake find a work around ? This seems to be a big cons for cmake, not for Hypre. The thing I do not get, is that, for instance, VisIt uses cmake while Silo (which VisIt depends on) uses configure : and that works. Sorry, my knowledge of cmake is tiny...

Not in general, every consumer could write their own find_package module, or HYPRE's autotools install could install one for the consumer, or one could contribute a module to CMake upstream, or one could use a pkg-config file instead if hypre provides one.

File: var/spack/repos/builtin/packages/hypre/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/hypre/package.py:172]: There was a typo in the title (`2.21.1` rather than `2.22.1` so I wanted to check if in fact in this case you meant `@2.21.0:`

**Review by scheibelp — COMMENTED** [review]
This looks generally good but I wanted to check: do earlier versions of `hypre` also use Autotools? Or is it just starting with `2.22.1`?

File: var/spack/repos/builtin/packages/hypre/package.py
Comment (cessenat) [reply var/spack/repos/builtin/packages/hypre/package.py:172]: Thanks for correcting the typo. The comment applies to 2.21.0 so it's OK.

**Comment by cessenat** [comment]
> This looks generally good but I wanted to check: do earlier versions of `hypre` also use Autotools? Or is it just starting with `2.22.1`?

I checked there has always been a CMakeLists.txt in Hypre, at least the first release declared for spack.
This was for Windows compilation with Visual Studio.
I checked randomly release 2.14.0 and both cmake and configure worked well. It took longer to compile with configure and the final static lib is bigger.  I'm sorry but I did not check thir compared functionalities.

**Comment by cessenat** [comment]
> > 
> > Does it mean that cmake does not work on any package that has been installed by configure ? Can't cmake find a work around ? This seems to be a big cons for cmake, not for Hypre. The thing I do not get, is that, for instance, VisIt uses cmake while Silo (which VisIt depends on) uses configure : and that works. Sorry, my knowledge of cmake is tiny...
> 
> Not in general, every consumer could write their own find_package module, or HYPRE's autotools install could install one for the consumer, or one could contribute a module to CMake upstream, or one could use a pkg-config file instead if hypre provides one.

OK so if I'm correct there are other ways to achieve what you need than using CMake to compile Hypre.
But I can imagine it is the best solution when one needs CMake for a package that depends on Hypre.
This is why I hope we could manage to keep both Autotools and CMake, at least for some time.
Well for now it is a "Package", wrongly agnostic because this is indeed a hidden Autotools one.
But maybe I should submit a hypre-autotools package and let people decide which package people want to use for Hypre in spack - which definitely should not be the general "Package".

File: var/spack/repos/builtin/packages/hypre/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/hypre/package.py]: Sorry I forgot to ask earlier: if the below code is supposed to replace `--enable-cub`, shouldn't we only add that if `spec.satisfies('@:2.20.99')` (i.e. for all versions before `2.21.0`). For example this could be removed and we could add to the below if statement:

```
if '@2.21.0:' in self.spec:
    ...
else:
    configure_args.append('--enable-cub')
```

**Comment by junghans** [comment]
@scheibelp what are we doing about cmake?

File: var/spack/repos/builtin/packages/hypre/package.py
Comment (cessenat) [reply var/spack/repos/builtin/packages/hypre/package.py]: That is right, so I modified accordingly.
Further, to show the benefit of switching from "Package" to "AutotoolsPackage", I used the with_or_without and enable_or_disable condensed forms. Of course switching to cmake one day is another matter.

**Comment by cessenat** [comment]
BTW using Autotools lets us have the standard stages : configure/build/install instead of only one big stage 'install' as is the case now.

**Comment by osborn9** [comment]
> > This looks generally good but I wanted to check: do earlier versions of `hypre` also use Autotools? Or is it just starting with `2.22.1`?
> 
> I checked there has always been a CMakeLists.txt in Hypre, at least the first release declared for spack.
> This was for Windows compilation with Visual Studio.
> I checked randomly release 2.14.0 and both cmake and configure worked well. It took longer to compile with configure and the final static lib is bigger. I'm sorry but I did not check thir compared functionalities.

Yes, autotools / configure has been the primary mode for hypre. CMake support is catching up, but not all features are exposed via CMake yet. For instance, CUDA support via CMake was only added in v2.21.0 (and not all options are configurable via CMake).

**Comment by osborn9** [comment]
> @scheibelp what are we doing about cmake?

Is there a way to use the suggestion from @cessenat https://github.com/spack/spack/pull/24817#issuecomment-912263550 so that we can have both options within spack? @junghans

**Comment by junghans** [comment]
> > @scheibelp what are we doing about cmake?
> 
> Is there a way to use the suggestion from @cessenat [#24817 (comment)](https://github.com/spack/spack/pull/24817#issuecomment-912263550) so that we can have both options within spack? @junghans

I am happy as long as I don't have to write that spackage ;-) But yeah, honestly we need `hypre[-+]cmake` to bump the version of `cabana`.

**Review by osborn9 — DISMISSED** [review]
Thanks for these updates @cessenat!

**Comment by scheibelp** [comment]
@spackbot rerun pipeline

**Comment by scheibelp** [comment]
@spackbot re-run pipeline

**Comment by scheibelp** [comment]
> @scheibelp what are we doing about cmake?

To address this explicitly: I think the best approach for now is to merge this, then add CMake support later (since it sounds like CMake doesn't have the same options). We are working on how to handle packages with support for multiple build systems so that should become easier in the future. For CMake packages which depend on this, hopefully there are associated CMake variables they can set to point them to this installation.

Does that sound reasonable?

**Comment by osborn9** [comment]
> @scheibelp what are we doing about cmake?

@cessenat @junghans As suggested in 

> > @scheibelp what are we doing about cmake?
> 
> To address this explicitly: I think the best approach for now is to merge this, then add CMake support later (since it sounds like CMake doesn't have the same options). We are working on how to handle packages with support for multiple build systems so that should become easier in the future. For CMake packages which depend on this, hopefully there are associated CMake variables they can set to point them to this installation.
> 
> Does that sound reasonable?

From the hypre perspective, this is good as this PR is just an expansion of our existing spack package. As mentioned in #https://github.com/spack/spack/pull/24817#issuecomment-915689257, the hypre dev team will make a decision on how to proceed regarding CMake.

**Review by ulrikeyang — DISMISSED** [review]
This looks good to me. Thank you.

**Comment by scheibelp** [comment]
@spackbot re-run pipeline

**Comment by junghans** [comment]
> > @scheibelp what are we doing about cmake?
> 
> To address this explicitly: I think the best approach for now is to merge this, then add CMake support later (since it sounds like CMake doesn't have the same options). We are working on how to handle packages with support for multiple build systems so that should become easier in the future. For CMake packages which depend on this, hopefully there are associated CMake variables they can set to point them to this installation.
> 
> Does that sound reasonable?

That won't work, because the Autotools version doesn't install a CMake module at all and hence there are also no variables to pick up. In #25847 we will depend on that, maybe we could install a custom FindHYPRE.cmake for the time being as a workaround.

**Comment by scheibelp** [comment]
> That won't work, because the Autotools version doesn't install a CMake module at all and hence there are also no variables to pick up. In #25847 we will depend on that

Ah OK, sometimes in CMake the dependent will have an option like `-DHYPRE_DIR` (not exactly like that but hopefully that conveys the idea).

> maybe we could install a custom FindHYPRE.cmake for the time being as a workaround.

That would also work in the meantime, although I agree it would be ideal to update Spack to make this easier (which we are working on).

**Comment by scheibelp** [comment]
@spackbot re-run pipeline

**Comment by scheibelp** [comment]
@spackbot run pipeline

**Comment by scheibelp** [comment]
@spackbot re-run pipeline

**Review by balay — DISMISSED** [review]

**Comment by scheibelp** [comment]
It looks like `mfem` is failing to build with this version of `hypre`, does the error message at https://gitlab.spack.io/spack/spack/-/jobs/741884 provide any clues to you on whether to update constraints in `mfem`?

We can also adjust `share/spack//gitlab/cloud_pipelines/stacks/radiuss/spack.yaml` to build `mfem` with a different version of `hypre`, but if `mfem` should build with an earlier version in general, then we should probably update the `mfem` package.

**Comment by scheibelp** [comment]
From that link, the error message suggests to me that the latest version of `mfem` (in Spack) does not work with this version of `hypre`:

```
1910  >> 164    make: *** [linalg/hypre_parcsr.o] Error 1
1911     165    make: *** Waiting for unfinished jobs....
1912     166    linalg/complex_operator.cpp: In member function 'mfem::HypreParMatr
1913            ix* mfem::ComplexHypreParMatrix::GetSystemMatrix() const':
1914  >> 167    linalg/complex_operator.cpp:792:4: error: 'hypre_ParCSRMatrixSetRow
1915            StartsOwner' was not declared in this scope
1916     168        hypre_ParCSRMatrixSetRowStartsOwner(hA,1);
1917     169        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by balay** [comment]
yeah `hypre@2.22.1` has some changes that break current releases of mfem, sundials, petsc

**Comment by balay** [comment]
I think the following change for petsc
```diff
diff --git a/var/spack/repos/builtin/packages/petsc/package.py b/var/spack/repos/builtin/packages/petsc/package.py
index 683420b883..0821016fc6 100644
--- a/var/spack/repos/builtin/packages/petsc/package.py
+++ b/var/spack/repos/builtin/packages/petsc/package.py
@@ -250,8 +250,10 @@ class Petsc(Package, CudaPackage, ROCmPackage):
     depends_on('hypre@:2.13.99+mpi~internal-superlu+int64', when='@:3.8.99+hypre+mpi~complex+int64')
     depends_on('hypre@2.14:2.18.2+mpi~internal-superlu~int64', when='@3.9:3.13.99+hypre+mpi~complex~int64')
     depends_on('hypre@2.14:2.18.2+mpi~internal-superlu+int64', when='@3.9:3.13.99+hypre+mpi~complex+int64')
-    depends_on('hypre@2.14:+mpi~internal-superlu~int64', when='@3.14:+hypre+mpi~complex~int64')
-    depends_on('hypre@2.14:+mpi~internal-superlu+int64', when='@3.14:+hypre+mpi~complex+int64')
+    depends_on('hypre@2.14:2.22.0+mpi~internal-superlu~int64', when='@3.14:3.15.99+hypre+mpi~complex~int64')
+    depends_on('hypre@2.14:2.22.0+mpi~internal-superlu+int64', when='@3.14:3.15.99+hypre+mpi~complex+int64')
+    depends_on('hypre@2.14:+mpi~internal-superlu~int64', when='@3.16:+hypre+mpi~complex~int64')
+    depends_on('hypre@2.14:+mpi~internal-superlu+int64', when='@3.16:+hypre+mpi~complex+int64')
     depends_on('hypre@xsdk-0.2.0+mpi~internal-superlu+int64', when='@xsdk-0.2.0+hypre+mpi~complex+int64')
     depends_on('hypre@xsdk-0.2.0+mpi~internal-superlu~int64', when='@xsdk-0.2.0+hypre+mpi~complex~int64')
     depends_on('hypre@develop+mpi~internal-superlu+int64', when='@main+hypre+mpi~complex+int64')
```

**Comment by scheibelp** [comment]
@cessenat would you be able to add these constraints to the mentioned packages? e.g. for `mfem` I think the diff would be:

```
diff --git a/var/spack/repos/builtin/packages/mfem/package.py b/var/spack/repos/builtin/packages/mfem/package.py
index ecd470c4cb..cc8ce1eb13 100644
--- a/var/spack/repos/builtin/packages/mfem/package.py
+++ b/var/spack/repos/builtin/packages/mfem/package.py
@@ -199,8 +199,7 @@ class Mfem(Package, CudaPackage, ROCmPackage):
 
     depends_on('mpi', when='+mpi')
     depends_on('hypre@2.10.0:2.13.99', when='@:3.3.99+mpi')
-    depends_on('hypre@:2.20.0', when='@3.4:4.2.99+mpi')
-    depends_on('hypre', when='@4.3.0:+mpi')
+    depends_on('hypre@:2.20.0', when='@3.4:4.3.0+mpi')
 
     depends_on('metis', when='+metis')
     depends_on('blas', when='+lapack')
```

**Comment by cessenat** [comment]
Thank you @scheibelp and @balay for your suggestions to make MFEM and PETs compatible.
I made a separate PR  for that https://github.com/spack/spack/pull/25902
I made another separate PR for sundials https://github.com/spack/spack/pull/25903
As @scheibelp you may notice I changed your suggestion a little to reflect the fact that MFEM 4.3.0 is fine with Hypre 2.22.0.

**Comment by scheibelp** [comment]
> Thank you @scheibelp and @balay for your suggestions to make MFEM and PETs compatible.
> I made a separate PR for that #25902
> I made another separate PR for sundials #25903

Sorry for getting back to you late on this, but I think you'll need to put all those changes in this PR, otherwise the CI will fail (e.g. https://gitlab.spack.io/spack/spack/-/jobs/741884). I think overall that is good (since it ensures that users wouldn't ever be stuck between this commit and the commit which updates `mfem` constraints).

**Comment by cessenat** [comment]
Closing to re-run pipeline in case sundials is not run.

**Comment by cessenat** [comment]
Now pipeline should work since both sundials and mfem/petsc should work.
There is PR https://github.com/spack/spack/pull/26212 from @tldahlgren, which I rebased this PR on.
Unfortunately it seems to have dismissed reviews from @balay @ulrikeyang and @osborn9 despite just adding 
`     tags     = ['radiuss']`
I think all conditions are met to merge, do you think so @scheibelp ?

**Review by scheibelp — APPROVED** [review]

**Comment by cessenat** [comment]
Thanks everybody !

**Pull request merged**
